src/vocabulary.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::BTreeMap;
+
+use mentat_core::{
+    Attribute,
+    Entid,
+    HasSchema,
+    KnownEntid,
+    NamespacedKeyword,
+    TypedValue,
+    ValueType,
+};
+
+pub use mentat_core::attribute;
+use mentat_core::attribute::Unique;
+
+use ::{
+    CORE_SCHEMA_VERSION,
+    IntoResult,
+};
+
+use ::conn::{
+    InProgress,
+    Queryable,
+};
+
+use ::errors::{
+    ErrorKind,
+    Result,
+};
+
+use ::entity_builder::{
+    BuildTerms,
+    TermBuilder,
+    Terms,
+};
+
+pub use mentat_db::AttributeBuilder;
+
+pub type Version = u32;
+pub type Datom = (Entid, Entid, TypedValue);
+
+/// `Attribute` instances not only aren't named, but don't even have entids.
+/// We need two kinds of structure here: an abstract definition of a vocabulary in terms of names,
+/// and a concrete instance of a vocabulary in a particular store.
+/// Note that, because it's possible to 'flesh out' a vocabulary with attributes without bumping
+/// its version number, we need to know the attributes that the application cares about -- it's
+/// not enough to know the name and version. Indeed, we even care about the details of each attribute,
+/// because that's how we'll detect errors.
+#[derive(Debug)]
+pub struct Definition {
+    pub name: NamespacedKeyword,
+    pub version: Version,
+    pub attributes: Vec<(NamespacedKeyword, Attribute)>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Vocabulary {
+    pub entity: Entid,
+    pub version: Version,
+    attributes: Vec<(Entid, Attribute)>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct Vocabularies(pub BTreeMap<NamespacedKeyword, Vocabulary>);   // N.B., this has a copy of the attributes in Schema!
+
+impl Vocabularies {
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn get(&self, name: &NamespacedKeyword) -> Option<&Vocabulary> {
+        self.0.get(name)
+    }
+}
+
+lazy_static! {
```

Comment:
What's wrong with the copy in `core/src/values.rs`?

Reply:
Those are `edn::Value`s, not `NamespacedKeyword`s. `edn::Value` isn't even guaranteed to be a keyword, even ignoring the dependency on `edn`.

In the future I'd like to do two things:

- Disentangle the EDN stuff as much as possible. If this takes flight, it will likely use CBOR and JSON, not (just) EDN.
- Rethink `NamespacedKeyword` to use some kind of borrow/reference string wrapper, and also do away with the prefix/suffix pairing.